### PR TITLE
Modernizing CMake and packaging

### PR DESCRIPTION
**Pull request by dssgabriel**
We should use more modern constructs, such as `target_sources` and `FILE_SET`s, to improve packaging of KokkosComm and avoid having to deal with `{BUILD,INSTALL}_INTERFACE` generator expressions.

This PR updates all of our CMake files to this end, and also brings some improvements:
- bump required CMake version to 3.23 (available on latest Ubuntu LTS) 
- set the project language to `CXX`;
- fix SemVer number (`0.0.2` -> `0.2.0`);
- add project descriptions and homepage URL;
- clean up things which aren't needed/used.

I use the following resources for updating our approach:
- [Modern CMake Packaging guide](https://blog.nickelp.ro/posts/cmake-pkg/) (goes into more detail as to why `FILE_SET`s are the preferred approach for packaging with CMake)
- [`target_sources` and `FILE_SET`s documentation](https://cmake.org/cmake/help/latest/command/target_sources.html)

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -1,108 +1,85 @@
...
-
-
-if(KOKKOSCOMM_ENABLE_TESTS)
-    enable_testing()
-    add_subdirectory(unit_tests)
-endif()
-if(KOKKOSCOMM_ENABLE_PERFTESTS)
-    enable_testing()
-    add_subdirectory(perf_tests)
-endif()
```
Comment (cwpearson) [inline CMakeLists.txt:108]: Why does this get removed?

File: CMakeLists.txt [hunk 2]
```diff
@@ -1,108 +1,85 @@
...
+#
+#@HEADER
+
 # 3.9: MPI::MPI_CXX
 # 3.12: CMAKE_PROJECT_VERSION_MAJOR
 # 3.15: PUBLIC_HEADER files for interface libraries
-cmake_minimum_required(VERSION 3.12)
-project(KokkosComm VERSION 0.0.2)
+# 3.23: FILE_SETs for easy export of header-only libraries
+cmake_minimum_required(VERSION 3.23)
```
Comment (cwpearson) [inline CMakeLists.txt:21]: Kokkos only requires 3.16, but new CMake is easy enough to get so I think this is okay.

**Comment by cwpearson** [comment]
Not sure why the CI isn't running

File: CMakeLists.txt (hunk 1 above)
Comment (dssgabriel) [reply CMakeLists.txt:108]: I'm sorry, I moved it to lines 47-52 so it's next to the `add_subdirectory(src)`. I can place it back here if it makes more sense.

**Comment by dssgabriel** [comment]
Not sure either :confused:
CI took a very long time to start on my two previous pushes as well.

File: CMakeLists.txt (hunk 1 above)
Comment (cwpearson) [reply CMakeLists.txt:108]: Thanks, I just missed it. Either one is fine with me.

**Review by cwpearson — DISMISSED** [review]

**Comment by dssgabriel** [comment]
It appears that the Ubuntu 22.04 LTS image provided by NVIDIA doesn't ship with a recent enough CMake. I don't know if we can workaround this in our CI?

**Comment by dssgabriel** [comment]
I am not sure why the CMake version in the `CompileOnly` test is outdated but not in the other tests...
The `nvidia/cuda` container we pull from is [based on Ubuntu 22.04](https://gitlab.com/nvidia/container-images/cuda/-/blob/master/dist/12.5.0/ubuntu2204/base/Dockerfile#L1) and GitHub's `ubuntu-latest` runner image has CMake 3.30 (as documented [here](https://github.com/actions/runner-images/blob/main/images/ubuntu/Ubuntu2204-Readme.md#tools)).

**Comment by cwpearson** [comment]
I can just make a PR that installs 3.23.0 in this container.

**Comment by cwpearson** [comment]
https://github.com/kokkos/kokkos-comm/pull/124

**Comment by dssgabriel** [comment]
Thank you Carl! Can we manually restart the failing CI test without committing anything?

**Comment by cwpearson** [comment]
I'm not sure, but I think you have to rebase your PR on 124 before the test will be changed. You can add the changes that we made in 124 to your PR if you like.

File: CMakeLists.txt [hunk 3]
```diff
@@ -1,108 +1,85 @@
...
-    COMPATIBILITY SameMajorVersion
-)
-
-export (
-    TARGETS KokkosComm
-    NAMESPACE "${PROJECT_NAME}::"
-    FILE ${PROJECT_BINARY_DIR}/cmake/KokkosComm/KokkosCommTargets.cmake
+  ${PROJECT_BINARY_DIR}/cmake/${PROJECT_NAME}/${PROJECT_NAME}Version.cmake
+  COMPATIBILITY SameMajorVersion
+  ARCH_INDEPENDENT # KokkosComm is a header-only library
```
Comment (cedricchevalier19) [inline CMakeLists.txt]: Is using `${PROJECT_NAME}` instead of hardcoding `KokkosComm` really useful?

File: cmake/flags.cmake [hunk 4]
```diff
@@ -20,47 +36,52 @@ check_cxx_compiler_flag(-Wformat=2 CXX_HAS_WFORMAT2)
...
+    )
+  endif()
+  # gtest includes sometimes yield this warning
+  if(CXX_HAS_NO_GNU_ZERO_VARIADIC_MACRO_ARGUMENTS)
+    target_compile_options(
+      ${ADD_CXX_FLAGS_TARGET}
+      ${TARGET_COMPILE_OPTIONS_KEYWORD}
+      $<BUILD_INTERFACE:-Wno-gnu-zero-variadic-macro-arguments>
+    )
+  endif()
```
Comment (cedricchevalier19) [inline cmake/flags.cmake]: I favor removing this function and creating a target interface that provides the correct options. You can take the stuff I have done in #12.

File: src/KokkosComm/CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+#
+# Part of Kokkos, under the Apache License v2.0 with LLVM Exceptions.
+# See https://kokkos.org/LICENSE for license information.
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+#
+#@HEADER
+
+set(TARGET_LIBRARY kokkoscomm)
+
+include(${PROJECT_SOURCE_DIR}/cmake/flags.cmake)
```
Comment (cedricchevalier19) [inline src/KokkosComm/CMakeLists.txt]: If you change the flags to a target, you do not need to include this anymore.

File: src/KokkosComm/CMakeLists.txt [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+  @ONLY
+)
+target_sources(
+  KokkosComm
+  INTERFACE
+    FILE_SET kokkoscomm_config_headers
+    TYPE HEADERS
+    BASE_DIRS ${CMAKE_BINARY_DIR}/src
+    FILES ${PROJECT_BINARY_DIR}/src/KokkosComm/config.hpp
+)
```
Comment (cedricchevalier19) [inline src/KokkosComm/CMakeLists.txt]: I favor putting this code close to the top of the file (or even higher in the directory hierarchy).

File: CMakeLists.txt [hunk 7]
```diff
@@ -1,108 +1,85 @@
...
+include(CMakePackageConfigHelpers)
 
-target_include_directories(KokkosComm INTERFACE
-    $<BUILD_INTERFACE:${CMAKE_CURRENT_BINARY_DIR}/src/>
-    $<INSTALL_INTERFACE:${CMAKE_INSTALL_INCLUDEDIR}>
+# Generate package config file
+configure_package_config_file(
+  ${PROJECT_SOURCE_DIR}/cmake/Config.cmake.in
+  ${PROJECT_BINARY_DIR}/cmake/${PROJECT_NAME}/${PROJECT_NAME}Config.cmake
+  INSTALL_DESTINATION ${CMAKE_INSTALL_LIBDIR}/cmake/${PROJECT_NAME}
```
Comment (cedricchevalier19) [inline CMakeLists.txt]: Is `${CMAKE_INSTALL_LIBDIR}` correct now that `GNUInstallDirs` is not included anymore?

**Review by cedricchevalier19 — COMMENTED** [review]
Some of the changes are pure formatting. Can you add a linter in the CI, or can you defer these changes to another PR?

File: CMakeLists.txt (hunk 3 above)
Comment (dssgabriel) [reply CMakeLists.txt]: Done across all CMake files in [a0a34db](https://github.com/kokkos/kokkos-comm/pull/122/commits/a0a34dbc14c4863ed12b0e594cc73494176d5631).

File: cmake/flags.cmake (hunk 4 above)
Comment (dssgabriel) [reply cmake/flags.cmake]: File removed in favor of what you had in #12 (see [8ce3e76](https://github.com/kokkos/kokkos-comm/pull/122/commits/8ce3e760c7a951e1a09cead02feb7124997990d0) and [22bd01f](https://github.com/kokkos/kokkos-comm/pull/122/commits/22bd01f5466c37671b015ce7f930dae170631647)).

File: src/KokkosComm/CMakeLists.txt (hunk 5 above)
Comment (dssgabriel) [reply src/KokkosComm/CMakeLists.txt]: Done in [22bd01f](https://github.com/kokkos/kokkos-comm/pull/122/commits/22bd01f5466c37671b015ce7f930dae170631647).

File: src/KokkosComm/CMakeLists.txt (hunk 6 above)
Comment (dssgabriel) [reply src/KokkosComm/CMakeLists.txt]: Moved the config header generation to the top-level CMake file in [4d7526c](https://github.com/kokkos/kokkos-comm/pull/122/commits/4d7526ced834a99a3a16e98680154aa189800204).
I left the `target_sources` here though, as it is directly related to the `KokkosComm` library target that is defined in this file.

File: CMakeLists.txt (hunk 7 above)
Comment (dssgabriel) [reply CMakeLists.txt]: Indeed, I had missed that, thank you! Added it back as part of [4d7526c](https://github.com/kokkos/kokkos-comm/pull/122/commits/4d7526ced834a99a3a16e98680154aa189800204).

**Review by cwpearson — APPROVED** [review]

**Pull request merged**